### ISSUE TITLE
Cannot fullgraph differentiate through boolean masking

### ISSUE THREAD

**Issue by ezyang** [pytorch/pytorch#114123]
### 🐛 Describe the bug

When you turn on capture_dynamic_output_ops this preexisting test starts failing

```
    def test_setitem_boolean_mask_diff(self):
        def fn(x, b, y):
            x = x.clone()
            x[b] = y
            return x
        
        opt_fn = torch._dynamo.optimize("aot_eager")(fn)
        x = torch.randn(4, requires_grad=True)
        b = torch.tensor([True, False, True, False])
        y = torch.randn(2, requires_grad=True)
        opt_fn(x, b, y)
```

The failure is

```
torch._dynamo.exc.BackendCompilerFailed: backend='compiler_fn' raised:
GuardOnDataDependentSymNode: It appears that you're trying to get a value out of symbolic int/float whose value is data-dependent (and thus we do not know the true value.)  The expression we were trying to evaluate is Eq(i0, 2) (unhinted: Eq(i0, 2)).  Scroll up to see where each of these data-dependent accesses originally occurred.
```

Relevant C++ stacktrace:

```
[2023-11-20 06:29:53,710] [0/0] torch.fx.experimental.symbolic_shapes: [DEBUG]   File "??", line 0, in torch::autograd::validate_outputs(std::vector<torch::autograd::Edge, std::allocator<torch::autograd::Edge> > const&, std::vector<at::Tensor, std::allocator<at::Tensor> >&, std::function<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > (std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)> const&)
[2023-11-20 06:29:53,710] [0/0] torch.fx.experimental.symbolic_s[2023-11-20 06:29:53,710] [0/0] torch.fx.experimental.symbolic_shapes: [DEBUG]   File "??", line 0, in torch::autograd::validate_outputs(std::vector<torch::autograd::Edge, std::allocator<torch::autograd::Edge> > const&, std::vector<at::Tensor, std::allocator<at::Tensor> >&, std::function<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > (std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)> const&)
hapes: [DEBUG]   File "??", line 0, in torch::autograd::InputMetadata::is_same_shape(at::Tensor const&) const
[2023-11-20 06:29:53,710] [0/0] torch.fx.experimental.symbolic_shapes: [DEBUG]   File "??", line 0, in c10::SymBool::guard_bool(char const*, long) const
[2023-11-20 06:29:53,710] [0/0] torch.fx.experimental.symbolic_shapes: [DEBUG]   File "??", line 0, in torch::autograd::InputMetadata::is_same_shape(at::Tensor const&) const
[2023-11-20 06:29:53,710] [0/0] torch.fx.experimental.symbolic_shapes: [DEBUG]   File "??", line 0, in c10::SymBool::guard_bool(char const*, long) const
```

The relevant formula is `Error detected in IndexPutBackward0`

### Versions

main

cc @msaroufim @wconstab @bdhirsh @anijain2305 @zou3519

**Comment by masnesral** [comment]
Looks like this might have been fixed. @ezyang, mind double-checking my work:

```
(pytorch-3.10_3)  (viable/strict) ~/local/pytorch-3.10_3 $ git diff
diff --git a/test/dynamo/test_repros.py b/test/dynamo/test_repros.py
index da11dd36bdb..ecf0ab9f1b7 100644
--- a/test/dynamo/test_repros.py
+++ b/test/dynamo/test_repros.py
@@ -49,6 +49,7 @@ from torch.testing._internal.common_utils import (
 )
 from torch.testing._internal.two_tensor import TwoTensor

+torch._dynamo.config.capture_dynamic_output_shape_ops = True

 _orig_module_call = torch.nn.Module.__call__

(pytorch-3.10_3)  (viable/strict) ~/local/pytorch-3.10_3 $ python test/dynamo/test_repros.py -k boolean_mask_diff
frames [('total', 2), ('ok', 2)]
unimplemented []
graph_break [('boolean masking setitem backwards, see https://github.com/pytorch/pytorch/issues/114123', 1)]
stats [('calls_captured', 1), ('unique_graphs', 1)]
aot_autograd [('total', 1), ('ok', 1)]
.frames [('total', 2), ('ok', 2)]
unimplemented []
graph_break [('boolean masking setitem backwards, see https://github.com/pytorch/pytorch/issues/114123', 1)]
stats [('calls_captured', 1), ('unique_graphs', 1)]
aot_autograd [('total', 1), ('ok', 1)]
.
----------------------------------------------------------------------
Ran 2 tests in 3.092s

OK
```
